src/sparseml/pytorch/image_classification/utils/trainer.py

```diff
@@ -0,0 +1,284 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Trainers for image classification.
+"""
+
+import logging
+from abc import ABC, abstractmethod
+from typing import Any, Callable, Dict, List, Optional
+
+import torch
+from torch.utils.data import DataLoader
+
+from sparseml.pytorch.optim import ScheduledModifierManager, ScheduledOptimizer
+from sparseml.pytorch.utils import (
+    DEFAULT_LOSS_KEY,
+    CrossEntropyLossWrapper,
+    ModuleDeviceContext,
+    ModuleTester,
+    ModuleTrainer,
+    default_device,
+)
+
+
+_LOGGER = logging.getLogger(__file__)
+
+__all__ = [
+    "ImageClassificationTrainer",
+]
+
+
+class Trainer(ABC):
+    """
+    Abstract class for Trainers.
+    Creates a contract that all trainers must have a run_one_epoch method.
+    """
+
+    @abstractmethod
+    def run_one_epoch(self):
+        """
+        Runs one epoch of training.
+        """
+        raise NotImplementedError
+
+
+class ImageClassificationTrainer(Trainer):
+    """
+    Trainer for image classification.
+
+    :param model: The loaded torch model to train.
+    :param key: The arch key of the model.
+    :param recipe_path: The path to the yaml file containing the modifiers and
+        schedule to apply them with; Can also provide a SparseZoo stub prefixed
+        with 'zoo:'.
+    :param ddp: bool indicating whether to use Distributed Data Parallel.
+    :param device: The device to train on. Defaults to torch default device.
+    :param use_mixed_precision: Whether to use mixed precision FP16 training.
+        Defaults to False.
+    :param val_loader: A DataLoader for validation data.
+    :param train_loader: A DataLoader for training data.
+    :param is_main_process: Whether the current process is the main process,
+        while training using DDP. Defaults to True.
+    :param loggers: A list of loggers to use during training process.
+    :param loss_fn: A Callabale loss function for training and validation
+        losses.
+    :param init_lr: The initial learning rate for the optimizer.Defaults to
+        1e-9
+    :param optim_name: str respresenting the optimizer type to use, one of
+        [SGD, Adam, RMSprop]. Defaults to `adam`.
+    :param optim_kwargs: dict of additional kwargs to pass to the optimizer.
+    """
+
+    def __init__(
+        self,
+        model: torch.nn.Module,
+        key: str,
+        recipe_path: str,
+        ddp: bool = False,
+        device: str = default_device(),
+        use_mixed_precision: bool = False,
+        val_loader: Optional[DataLoader] = None,
+        train_loader: Optional[DataLoader] = None,
+        is_main_process: bool = True,
+        loggers: Optional[List[Any]] = None,
+        loss_fn: Callable = lambda: CrossEntropyLossWrapper(),
+        init_lr=1e-9,
+        optim_name="adam",
```

Comment:
Nit: let's be consistent across docstrings and code: adam or Adam.

Reply:
Fixed

Reply:
Fixed in favor of "Adam"